File: khalti/src/main/java/com/khalti/form/Wallet/Wallet.java
package com.khalti.form.Wallet;

import android.Manifest;
import android.app.Dialog;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.khalti.R;
import com.khalti.SmsListener;
import com.khalti.carbonX.widget.Button;
import com.khalti.carbonX.widget.ExpandableLayout;
import com.khalti.carbonX.widget.FrameLayout;
import com.khalti.carbonX.widget.TextInputLayout;
import com.khalti.form.CheckOutActivity;
import com.khalti.utils.DataHolder;
import com.khalti.utils.Event;
import com.khalti.utils.RxBus;
import com.utila.AppPermissionUtil;
import com.utila.EmptyUtil;
import com.utila.NetworkUtil;
import com.utila.NumberUtil;
import com.utila.ResourceUtil;
import com.utila.UserInterfaceUtil;

import fontana.RobotoMediumTextView;
import rx.subscriptions.CompositeSubscription;


public class Wallet extends Fragment implements WalletContract.View {

    private TextInputEditText etMobile, etCode, etPIN;
    private TextInputLayout tilMobile, tilCode, tilPIN;
    private ExpandableLayout elConfirmation;
    private Button btnPay;
    private Dialog progressDialog;

    private FragmentActivity fragmentActivity;
    private WalletContract.Listener listener;
    private CompositeSubscription compositeSubscription;
    private SmsListener smsListener;


    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View mainView = inflater.inflate(R.layout.payment_form, container, false);
        fragmentActivity = getActivity();
        listener = new WalletPresenter(this);

        etMobile = mainView.findViewById(R.id.etMobile);
        etCode = mainView.findViewById(R.id.etCode);
        etPIN = mainView.findViewById(R.id.etPIN);
        tilMobile = mainView.findViewById(R.id.tilMobile);
        tilCode = mainView.findViewById(R.id.tilCode);
        tilPIN = mainView.findViewById(R.id.tilPIN);
        btnPay = mainView.findViewById(R.id.btnPay);
        elConfirmation = mainView.findViewById(R.id.elConfirmation);

        listener.setUpLayout();

        compositeSubscription = new CompositeSubscription();
        compositeSubscription.add(RxBus.getInstance().register(Event.class, event -> {
            if (event.getTag().equals("wallet_code")) {
                listener.setConfirmationCode(event);
            }
        }));

        return mainView;
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (EmptyUtil.isNotNull(compositeSubscription) && !compositeSubscription.isUnsubscribed()) {
            compositeSubscription.unsubscribe();
        }
        listener.toggleSmsListener(false);
    }

    @Override
    public void setEditTextListener() {
        if (EmptyUtil.isNotNull(etMobile)) {
            etMobile.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                    tilMobile.setErrorEnabled(false);
                    if (btnPay.getText().toString().toLowerCase().contains("confirm")) {
                        listener.toggleConfirmationLayout(false);
                    }
                }

                @Override
                public void afterTextChanged(Editable editable) {

                }
            });
        }

        if (EmptyUtil.isNotNull(etCode)) {
            etCode.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                    tilCode.setErrorEnabled(false);
                }

                @Override
                public void afterTextChanged(Editable editable) {

                }
            });
        }

        if (EmptyUtil.isNotNull(etPIN)) {
            etPIN.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                    tilPIN.setErrorEnabled(false);
                }

                @Override
                public void afterTextChanged(Editable editable) {

                }
            });
        }
    }

    @Override
    public void toggleProgressDialog(String action, boolean show) {
        FrameLayout flCircularProgress = (FrameLayout) fragmentActivity.getLayoutInflater().inflate(R.layout.component_circular_progress, null);
        String message = "";
        switch (action) {
            case "init":
                message = ResourceUtil.getString(fragmentActivity, R.string.init_payment);
                break;
            case "confirm":
                message = ResourceUtil.getString(fragmentActivity, R.string.confirming_payment);
                break;
        }

        if (show) {
            progressDialog = UserInterfaceUtil.runProgressDialog(fragmentActivity, message, ResourceUtil.getString(fragmentActivity, R.string.please_wait), flCircularProgress, () -> {
                listener.unSubscribe();
            });
        } else {
            if (EmptyUtil.isNotNull(progressDialog)) {
                progressDialog.dismiss();
            }
        }
    }

    @Override
    public void setEditTextError(String view, String error) {
        switch (view) {
            case "mobile":
                tilMobile.setError(error);
                break;
            case "code":
                tilCode.setError(error);
                break;
            case "pin":
                tilPIN.setError(error);
                break;
        }
    }

    @Override
    public void setButtonText(String text) {
        btnPay.setText(text);
    }

    @Override
    public void setButtonClickListener() {
        btnPay.setOnClickListener(view -> {
            if (btnPay.getText().toString().toLowerCase().contains("confirm")) {
                listener.confirmPayment(NetworkUtil.isNetworkAvailable(fragmentActivity), etCode.getText().toString(), etPIN.getText().toString());
            } else {
                if (AppPermissionUtil.checkAndroidPermission(fragmentActivity, Manifest.permission.RECEIVE_SMS)) {
                    listener.initiatePayment(NetworkUtil.isNetworkAvailable(fragmentActivity), etMobile.getText().toString());
                } else {
                    FrameLayout flPositive = (FrameLayout) fragmentActivity.getLayoutInflater().inflate(R.layout.component_flat_button, null);
                    RobotoMediumTextView tvPositive = flPositive.findViewById(R.id.tvButton);
                    tvPositive.setText(ResourceUtil.getString(fragmentActivity, R.string.allow));

                    FrameLayout flNegative = (FrameLayout) fragmentActivity.getLayoutInflater().inflate(R.layout.component_flat_button, null);
                    RobotoMediumTextView tvNegative = flNegative.findViewById(R.id.tvButton);
                    tvNegative.setText(ResourceUtil.getString(fragmentActivity, R.string.deny));

                    AppPermissionUtil.askPermission(fragmentActivity, Manifest.permission.RECEIVE_SMS, "Please allow permission to receive SMS", flPositive, flNegative, () ->
                            listener.initiatePayment(NetworkUtil.isNetworkAvailable(fragmentActivity), etMobile.getText().toString()));
                }
            }
        });
    }

    @Override
    public void setConfirmationCode(String code) {
        etCode.setText(code);
        etPIN.requestFocus();
    }

    @Override
    public void showNetworkError() {
        UserInterfaceUtil.showSnackBar(fragmentActivity, ((CheckOutActivity) this.fragmentActivity).cdlMain, ResourceUtil.getString(fragmentActivity, R.string.network_error_body),
                false, "", 0, 0, null);
    }

    @Override
    public void showMessageDialog(String title, String message) {
        FrameLayout flButton = (FrameLayout) fragmentActivity.getLayoutInflater().inflate(R.layout.component_flat_button, null);
        RobotoMediumTextView tvButton = flButton.findViewById(R.id.tvButton);
        tvButton.setText(ResourceUtil.getString(fragmentActivity, R.string.got_it));

        UserInterfaceUtil.showInfoDialog(fragmentActivity, title, message, true, true, flButton, new UserInterfaceUtil.DialogAction() {
            @Override
            public void onPositiveAction(Dialog dialog) {
                dialog.dismiss();
            }

            @Override
            public void onNegativeAction(Dialog dialog) {

            }
        });
    }

    @Override
    public void showInteractiveMessageDialog(String title, String message) {
        FrameLayout flPositive = (FrameLayout) fragmentActivity.getLayoutInflater().inflate(R.layout.component_flat_button, null);
        RobotoMediumTextView tvPositive = flPositive.findViewById(R.id.tvButton);
        tvPositive.setText(ResourceUtil.getString(fragmentActivity, R.string.ok));

        FrameLayout flNegative = (FrameLayout) fragmentActivity.getLayoutInflater().inflate(R.layout.component_flat_button, null);
        RobotoMediumTextView tvNegative = flNegative.findViewById(R.id.tvButton);
        tvNegative.setText(ResourceUtil.getString(fragmentActivity, R.string.cancel));

        UserInterfaceUtil.showInteractiveInfoDialog(fragmentActivity, title, message, true, true, flPositive, flNegative, new UserInterfaceUtil.DialogAction() {
            @Override
            public void onPositiveAction(Dialog dialog) {
                dialog.dismiss();
                listener.openKhaltiSettings();
            }

            @Override
            public void onNegativeAction(Dialog dialog) {
                dialog.dismiss();
            }
        });
    }

    @Override
    public String getStringFromResource(int id) {
        return ResourceUtil.getString(fragmentActivity, id);
    }

    @Override
    public void openKhaltiSettings() {
        Intent i;
        PackageManager manager = fragmentActivity.getPackageManager();
        try {
            i = manager.getLaunchIntentForPackage("com.khalti.red");
            if (i == null)
                throw new PackageManager.NameNotFoundException();
            i.addCategory(Intent.CATEGORY_LAUNCHER);
            startActivity(i);
        } catch (PackageManager.NameNotFoundException e) {
            listener.showMessageDialog("Error", ResourceUtil.getString(fragmentActivity, R.string.khalti_not_found));
        }
    }

    @Override
    public void closeWidget() {
        fragmentActivity.finish();
    }

    @Override
    public void toggleConfirmationLayout(boolean show) {
        String buttonText = show ? ResourceUtil.getString(fragmentActivity, R.string.confirm_payment) : "Pay Rs " + NumberUtil.convertToRupees(DataHolder.getConfig().getAmount());
        btnPay.setText(buttonText);
        etCode.setText("");
        etPIN.setText("");
        elConfirmation.toggleExpansion();
    }

    @Override
    public void toggleSmsListener(boolean listen) {
        if (listen) {
            IntentFilter intentFilter = new IntentFilter("android.provider.Telephony.SMS_RECEIVED");
            smsListener = new SmsListener();
            fragmentActivity.registerReceiver(smsListener, intentFilter);
        } else {
            if (EmptyUtil.isNotNull(smsListener)) {
                fragmentActivity.unregisterReceiver(smsListener);
            }
        }
    }

    @Override
    public void setListener(WalletContract.Listener listener) {
        this.listener = listener;
    }
}

File: khalti-android/src/main/java/com/khalti/checkOut/wallet/Wallet.kt
package com.khalti.checkOut.wallet

import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.Toast
import androidx.appcompat.app.AppCompatDialog
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textfield.TextInputLayout
import com.khalti.R
import com.khalti.signal.Signal
import com.khalti.utils.*
import kotlinx.android.synthetic.main.wallet_form.view.*
import java.util.*
import kotlin.math.abs

class Wallet : Fragment(), WalletContract.View {

    private var progressDialog: AppCompatDialog? = null
    private lateinit var fragmentActivity: FragmentActivity
    private lateinit var presenter: WalletContract.Presenter

    private var height: Int = 0
    private lateinit var mainView: View

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        mainView = inflater.inflate(R.layout.wallet_form, container, false)
        fragmentActivity = activity!!
        presenter = WalletPresenter(this)

        presenter.onCreate()

        return mainView
    }

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.onDestroy()
    }

    override val payButtonText: String get() = ViewUtil.getText(mainView.btnPay)

    override val formData: Map<String, String>
        get() = object : HashMap<String, String>() {
            init {
                put("mobile", ViewUtil.getText(mainView.etMobile))
                put("code", ViewUtil.getText(mainView.etCode))
                put("pin", ViewUtil.getText(mainView.etPIN))
            }
        }

    override fun toggleProgressDialog(action: String, show: Boolean): Signal<Boolean> {
        var signal = Signal<Boolean>()
        var message = ""
        when (action) {
            "init" -> message = ResourceUtil.getString(fragmentActivity, R.string.init_payment)
            "confirm" -> message = ResourceUtil.getString(fragmentActivity, R.string.confirming_payment)
        }

        if (show) {
            val pair = UserInterfaceUtil.runProgressDialog(fragmentActivity, message, ResourceUtil.getString(fragmentActivity, R.string.please_wait))
            progressDialog = pair.first
            signal = pair.second
        } else {
            progressDialog?.dismiss()
        }
        return signal
    }

    override fun toggleConfirmationLayout(show: Boolean) {
        val buttonText = if (show) ResourceUtil.getString(fragmentActivity, R.string.confirm_payment) else "Pay Rs ${NumberUtil.convertToRupees(Store.getConfig().amount)}"
        mainView.btnPay?.text = buttonText
        mainView.etCode?.setText("")
        mainView.etPIN?.setText("")
        mainView.elConfirmation.toggleExpansion()
    }

    override fun togglePinMessage(show: Boolean) {
        mainView.cvPinMessage?.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun setPinMessage(message: String) {
        mainView.tvPinMessage?.text = message + " " + ResourceUtil.getString(fragmentActivity, R.string.sms_info)
    }

    override fun setMobile(mobile: String) {
        mainView.etMobile?.setText(mobile)
        mainView.etMobile?.setSelection(mobile.length)
    }

    override fun setEditTextError(view: String, error: String?) {
        val isError = EmptyUtil.isNotNull(error)
        when (view) {
            "mobile" -> {
                mainView.tilMobile?.isErrorEnabled = isError
                mainView.tilMobile?.error = error
            }
            "code" -> {
                mainView.llCode.measure(0, 0)
                val beforeA = mainView.llCode.measuredHeight

                mainView.tilCode.isErrorEnabled = isError
                mainView.tilCode.error = error

                mainView.llCode.measure(0, 0)
                val afterA = mainView.llCode.measuredHeight

                height = abs(height + (afterA - beforeA))
            }
            "pin" -> {
                mainView.llPIN.measure(0, 0)
                val beforeB = mainView.llPIN.measuredHeight

                mainView.tilPIN.isErrorEnabled = isError
                mainView.tilPIN.error = error

                mainView.llPIN.measure(0, 0)
                val afterB = mainView.llPIN.measuredHeight

                height = abs(height + (afterB - beforeB))
            }
        }
    }

    override fun setButtonText(text: String) {
        mainView.btnPay?.text = text
    }

    override fun setConfirmationLayoutHeight(view: String) {
        val til: TextInputLayout
        val ll: LinearLayout
        if (view == "code") {
            til = mainView.tilCode
            ll = mainView.llCode
        } else {
            til = mainView.tilPIN
            ll = mainView.llPIN
        }
        if (EmptyUtil.isNotNull(til?.error)) {
            ll?.measure(0, 0)
            val beforeA = ll?.measuredHeight

            til?.setErrorEnabled(false)

            ll?.measure(0, 0)
            val afterA = ll?.measuredHeight

            if (EmptyUtil.isNotNull(beforeA) && EmptyUtil.isNotNull(afterA)) {
                height = abs(height + (beforeA!! - afterA!!))

                val layoutParams = mainView.llConfirmation.layoutParams as ExpandableLayout.LayoutParams
                layoutParams.height = mainView.llConfirmation.height - height
                mainView.llConfirmation.layoutParams = layoutParams

                height = 0
            }
        }
    }

    override fun setEditTextListener(): Map<String, Signal<CharSequence>> = object : HashMap<String, Signal<CharSequence>>() {
        init {
            put("mobile", ViewUtil.setTextChangeListener(mainView.etMobile))
            put("code", ViewUtil.setTextChangeListener(mainView.etCode))
            put("pin", ViewUtil.setTextChangeListener(mainView.etPIN))
        }
    }

    override fun setClickListener(): Map<String, Signal<Any>> = object : HashMap<String, Signal<Any>>() {
        init {
            put("pay", ViewUtil.setClickListener(mainView.btnPay))
            put("khalti", ViewUtil.setClickListener(mainView.ivKhalti))
        }
    }

    override fun showNetworkError() {
        UserInterfaceUtil.showSnackBar(fragmentActivity, (this.fragmentActivity as com.khalti.checkOut.CheckOutActivity).cdlMain, ResourceUtil.getString(fragmentActivity, R.string.network_error_body), null, Snackbar.LENGTH_LONG)
    }

    override fun showMessageDialog(title: String, message: String) {
        UserInterfaceUtil.showInfoDialog(fragmentActivity, title, message, autoDismiss = true, cancelable = true, positiveText = ResourceUtil.getString(fragmentActivity, R.string.got_it), negativeText = null)
    }

    override fun showPINDialog(title: String, message: String): Signal<Boolean> =
            UserInterfaceUtil.showInfoDialog(fragmentActivity, title, message, autoDismiss = true, cancelable = true, positiveText = ResourceUtil.getString(fragmentActivity, R.string.ok), negativeText = ResourceUtil.getString(fragmentActivity, R.string.cancel))

    override fun showSlogan() {
        Toast.makeText(fragmentActivity, ResourceUtil.getString(fragmentActivity, R.string.slogan), Toast.LENGTH_SHORT).show()
    }

    override fun showBranding() {
        mainView.llKhaltiBranding?.visibility = View.VISIBLE
    }

    override fun openKhaltiSettings() {
        var i: Intent?
        val manager = fragmentActivity.packageManager
        try {
            i = manager.getLaunchIntentForPackage("com.khalti.red")
            i = if (EmptyUtil.isNotNull(i)) i else manager.getLaunchIntentForPackage("com.khalti")
            if (EmptyUtil.isNull(i)) {
                throw PackageManager.NameNotFoundException()
            }
            i!!.addCategory(Intent.CATEGORY_LAUNCHER)

            val map = HashMap<String, String>()
            map["pin_settings"] = "pin_settings"
            i.putExtra("map", map)

            startActivity(i)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()

            /*presenter.showPINInBrowserDialog("Error", ResourceUtil.getString(fragmentActivity, R.string.khalti_not_found) + "\n\n" +
                    ResourceUtil.getString(fragmentActivity, R.string.set_pin_in_browser))*/
        }

    }

    override fun openLinkInBrowser(link: String) {
        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(link))
        startActivity(browserIntent)
    }

    override fun closeWidget() {
        fragmentActivity.finish()
    }

    override fun updateConfirmationHeight() {
        val layoutParams = mainView.llConfirmation.layoutParams as ExpandableLayout.LayoutParams
        layoutParams.height = mainView.llConfirmation.height + height
        mainView.llConfirmation.layoutParams = layoutParams

        height = 0
    }

    override fun getMessage(action: String): String {
        return when (action) {
            "pin_not_set" -> ResourceUtil.getString(fragmentActivity, R.string.pin_not_set)
            "pin_not_set_continue" -> ResourceUtil.getString(fragmentActivity, R.string.pin_not_set_continue)
            "khalti_not_found" -> ResourceUtil.getString(fragmentActivity, R.string.khalti_not_found)
            "set_pin_in_browser" -> ResourceUtil.getString(fragmentActivity, R.string.set_pin_in_browser)
            else -> ""
        }
    }

    override fun hasNetwork(): Boolean {
        return NetworkUtil.isNetworkAvailable(fragmentActivity)
    }

    override fun doesPackageExist(): Boolean {
        val manager = fragmentActivity.packageManager
        return try {
            var i = manager.getLaunchIntentForPackage("com.khalti.red")
            if (EmptyUtil.isNull(i)) {
                i = manager.getLaunchIntentForPackage("com.khalti")
            }

            EmptyUtil.isNotNull(i)

        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            false
        }
    }

    override fun setPresenter(presenter: WalletContract.Presenter) {
        this.presenter = presenter
    }
}